### PR TITLE
Add new APIs for recommendations summary at Clusters and Namespaces level

### LINKED ISSUE #892: Add new APIs to support cluster and namespace level recommendations
Need to add new APIs to provide cluster and namespace level of recommendations as part of LM demo

[Design document for reference](https://docs.google.com/document/d/1x0yqYsMhAgpUEARUp8-AQpYk_Lg7ik_J7wssVunYHMM/edit)

### PR DESCRIPTION
**Pull request by khansaad**
This PR fixes #892 

- New APIs - `/listClusters` and `/summarize` are added which provides summary of the recommendations at cluster and namespace level

### CODE REVIEW

File: src/main/java/com/autotune/analyzer/services/ListClusters.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Timer.Sample timerListExp = Timer.start(MetricsConfig.meterRegistry());
+        response.setStatus(HttpServletResponse.SC_OK);
+        response.setContentType(JSON_CONTENT_TYPE);
+        response.setCharacterEncoding(CHARACTER_ENCODING);
+        List<String> clusterNamesList;
+
+        // get cluster list from the DB and send JSON array as the response
+        try {
+            clusterNamesList = new ExperimentDBService().loadClusterNames();
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/services/ListClusters.java:55]: rename it to loadAllClusterNames

similar to what we have like loadAllResults or loadAllPerformanceProfiles  etc

File: src/main/java/com/autotune/analyzer/services/Summarize.java [hunk 2]
```diff
@@ -0,0 +1,915 @@
...
+    HashMap<String, SummarizeAPIObject> clusterSummaryCacheMap = new HashMap<>();
+    HashMap<String, SummarizeAPIObject> namespaceSummaryCacheMap = new HashMap<>();
+
+    @Override
+    public void init(ServletConfig config) throws ServletException {
+        super.init(config);
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/services/Summarize.java:61]: Add description here and for very function

File: src/main/java/com/autotune/analyzer/services/Summarize.java [hunk 3]
```diff
@@ -0,0 +1,915 @@
...
+            if (!KruizeSupportedTypes.SUMMARIZE_PARAMS_SUPPORTED.contains(param)) {
+                invalidParams.add(param);
+            }
+        }
+        if (invalidParams.isEmpty()) {
+            // Set default value if absent
+            if (summarizeType == null || summarizeType.isEmpty())
+                summarizeType = KruizeConstants.JSONKeys.CLUSTER;
+            // by default, db_flag will be false so the data will be fetched from cache only
+            if (fetchFromDB == null || fetchFromDB.isEmpty())
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/services/Summarize.java:84]: based on description should be renamed 
as fetchFromCache
fetchFrom(DB or Thanos) varaible can be used when we move to Thanos server
Just a thought so you can ignore this comment .

File: src/main/java/com/autotune/analyzer/services/Summarize.java [hunk 4]
```diff
@@ -0,0 +1,915 @@
...
+                fetchFromDB = AnalyzerConstants.BooleanString.FALSE;
+            if (isValidValue(summarizeType, fetchFromDB)) {
+                // load recommendations based on params
+                try {
+                    if (fetchFromDB.equals(AnalyzerConstants.BooleanString.TRUE)) {
+                        clusterSummaryCacheMap = new HashMap<>();
+                        namespaceSummaryCacheMap = new HashMap<>();
+                    }
+                    // get the clusters and namespaces
+                    HashMap<String, List<String>> namespacesWithClustersMap = new ExperimentDBService().loadNamespacesByClusterNames();
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/services/Summarize.java:94]: should be renamed as loadAllNamespaces

File: src/main/java/com/autotune/database/dao/ExperimentDAOImpl.java [hunk 5]
```diff
@@ -526,4 +528,186 @@ public List<KruizeResultsEntry> getKruizeResultsEntry(String experiment_name, Ti
...
+        }
+        return entries;
+    }
+
+    @Override
+    public HashMap<String, List<String>> loadNamespacesByClusterNames() throws Exception {
+        HashMap<String, List<String>> entries = new HashMap<>();
+        try (Session session = KruizeHibernateUtil.getSessionFactory().openSession()) {
+            EntityManager entityManager = session.getEntityManagerFactory().createEntityManager();
+            String sql = "SELECT cluster_name, JSON_AGG(DISTINCT extended_data->'kubernetes_objects'->0->'namespace') as namespaces " +
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/database/dao/ExperimentDAOImpl.java:695]: move query to string constant

File: src/main/java/com/autotune/analyzer/services/Summarize.java (hunk 4 above)
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/services/Summarize.java:94]: rename variable name as
ClusterNamespaceAssociationMap

and ExperimentDBService().loadAllClusterNamespaceAssociationMap

File: src/main/java/com/autotune/analyzer/services/Summarize.java [hunk 6]
```diff
@@ -0,0 +1,915 @@
...
+
+                    if (summarizeType.equalsIgnoreCase(KruizeConstants.JSONKeys.CLUSTER)) {
+                        // do summarization based on cluster
+                        summarizeBasedOnClusters(namespacesWithClustersMap, namespaceName, clusterName, summarizeAPIObjectList);
+                    } else {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/services/Summarize.java:99]: use else if(summarizeType.equalsIgnoreCase('nameSpace'){
}and else{
   not supported 
}

File: src/main/java/com/autotune/database/helper/DBConstants.java [hunk 7]
```diff
@@ -15,9 +17,13 @@ public static final class SQLQUERY {
         public static final String SELECT_FROM_RESULTS_BY_EXP_NAME_AND_MAX_END_TIME = String.format("from KruizeResultsEntry k WHERE k.experiment_name = :%s and k.interval_end_time = (SELECT MAX(e.interval_end_time) FROM KruizeResultsEntry e  where e.experiment_name = :%s )", KruizeConstants.JSONKeys.EXPERIMENT_NAME, KruizeConstants.JSONKeys.EXPERIMENT_NAME);
         public static final String SELECT_FROM_RECOMMENDATIONS_BY_EXP_NAME = "from KruizeRecommendationEntry k WHERE k.experiment_name = :experimentName";
         public static final String SELECT_FROM_RECOMMENDATIONS_BY_EXP_NAME_AND_END_TIME = String.format("from KruizeRecommendationEntry k WHERE k.experiment_name = :%s and k.interval_end_time= :%s", KruizeConstants.JSONKeys.EXPERIMENT_NAME, KruizeConstants.JSONKeys.INTERVAL_END_TIME);
+        public static final String SELECT_FROM_RECOMMENDATIONS_BY_CLUSTER_NAME = "from KruizeRecommendationEntry k WHERE k.cluster_name = :clusterName and k.experiment_name IN :experimentList ORDER by interval_end_time DESC LIMIT :limit";
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/database/helper/DBConstants.java:20]: Refrain from employing the 'limit' statement on its own...
Instead, employ the 'limit' statement in conjunction with carefully optimized 'where' conditions. The 'limit' statement processes the entire content and then returns a specified count. However, we can enhance efficiency by strategically utilizing 'where' conditions along with the 'limit' statement, thus mitigating the need to process the entire content.

File: src/main/java/com/autotune/analyzer/services/Summarize.java (hunk 2 above)
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/services/Summarize.java:73]: Done

File: src/main/java/com/autotune/database/dao/ExperimentDAOImpl.java (hunk 5 above)
Comment (khansaad) [reply src/main/java/com/autotune/database/dao/ExperimentDAOImpl.java]: Done

File: src/main/java/com/autotune/analyzer/services/Summarize.java (hunk 4 above)
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/services/Summarize.java]: Done

File: src/main/java/com/autotune/analyzer/services/Summarize.java (hunk 6 above)
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/services/Summarize.java:116]: Done

File: src/main/java/com/autotune/analyzer/serviceObjects/SummarizeAPIObject.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+import com.autotune.analyzer.recommendations.summary.ResourceInfo;
+import com.autotune.analyzer.recommendations.summary.Summary;
+import com.autotune.analyzer.utils.AnalyzerConstants;
+import com.autotune.utils.KruizeConstants;
+import com.google.gson.annotations.SerializedName;
+
+import java.util.HashMap;
+
+public class SummarizeAPIObject {
+
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/serviceObjects/SummarizeAPIObject.java:29]: Needs to have a API version similar to all the other APIs

**Comment by dinogun** [comment]
Closing stale PRs

**Pull request closed without merging**